corehq/apps/api/resources/v0_5.py

```diff
@@ -745,6 +747,74 @@ class Meta(CustomResourceMeta):
         paginator_class = DoesNothingPaginator
 
 
+class DataSourceConfigurationResource(CouchResourceMixin, HqBaseResource, DomainSpecificResourceMixin):
+    """
+    API resource for DataSourceConfigurations (UCR data sources)
+    """
+    id = fields.CharField(attribute='get_id', readonly=True, unique=True)
+    display_name = fields.CharField(attribute="display_name", null=True)
+    configured_filter = fields.DictField(attribute="configured_filter", use_in='detail')
+    configured_indicators = fields.ListField(attribute="configured_indicators", use_in='detail')
+
+    def _ensure_toggle_enabled(self, request):
+        if not toggles.USER_CONFIGURABLE_REPORTS.enabled_for_request(request):
+            raise ImmediateHttpResponse(HttpResponse(
+                json.dumps({"error": _("You don't have permission to access this API")}),
+                content_type="application/json",
+                status=401)
+            )
+
+    def obj_get(self, bundle, **kwargs):
+        self._ensure_toggle_enabled(bundle.request)
```

Comment:
I considered a number of possible options before landing on "call this method at the top of all the other methods". 

1. Override `dispatch()` and run the checks first: Doesn't work because the authorization checks which populate the user and domain happen inside there. 
2. Override `dispatch()` and run the checks second: Doesn't work since the update method will already have already run its course.

The other options that may be better, but just didn't seem worth it include:

1. Add a new Authentication class that can handle both a permission check and a feature flag check (or more generically, a combinable authentication class)
2. Add a new `flag_enabled` decorator for APIs that can expect a `bundle` instead of a `request` - seemed a little specific and potentially overkill.

Open to different opinions or ideas I missed though.

Reply:
Interesting, I see toggle checks and access to `request.user` in the dispatch method of `HqBaseResource`: https://github.com/dimagi/commcare-hq/blob/7c4b468e69f0bcdc174bad6f2388568f5ee3ef6e/corehq/apps/api/resources/__init__.py#L96-L109

Do those just not work then?

Fine to leave this as is.

Reply:
Yeah, it's a bit more complicated than I articulated (my comment was already long!).

Middleware runs prior to dispatch being called, so our `UsersMiddleware` will always populate an available domain, and if the user can be determined from the session then it will also be there: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/users/middleware.py#L37-L43

In the event of an alternative form of authentication - in the case I was testing, oauth, the user is populated via the `Authentication` class at this point in the superclass: https://github.com/django-tastypie/django-tastypie/blob/master/tastypie/resources.py#L495

Which eventually (in the case of oauth) finds its way to our decorator here: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/domain/decorators.py#L216

I didn't actually test API keys, but now I'm curious since it wasn't obvious at a glance where a user would be set.

**TL;DR: domain toggles should be fine. User toggles are more complicated around the edges and probably shouldn't be used in `dispatch`.**

Reply:
aw my links didn't do the fancy thing your links did. :disappointed:

Reply:
Thanks for the update, that's also what I surmised from a brief look at the code.

> aw my links didn't do the fancy thing your links did. disappointed

Interesting, I wonder why?

Reply:
FYI, API keys also work (only when used this way). The user gets set here: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/domain/auth.py#L291-L292

Reply:
> Interesting, I wonder why?

It looks like it doesn't expand the links if it's to a branch - presumably because it could change at any time.